### ISSUE TITLE
DeepLab v3+ Not Supported

### ISSUE THREAD

**Issue by ChiFang** [tensorflow/tensorflow#17468]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**:
binary: by "pip3 install --upgrade tensorflow-gpu"
- **TensorFlow version (use command below)**:
tensorflow-gpu (1.6.0)
- **Python version**: 
Python 3.5.2 
- **Bazel version (if compiling from source)**:  0.11.0
- **GCC/Compiler version (if compiling from source)**:
gcc (Ubuntu 5.4.0-6ubuntu1~16.04.9) 5.4.0 20160609
- **CUDA/cuDNN version**:
CUDA 9.0 cuDNN 7
- **GPU model and memory**: Null
- **Exact command to reproduce**:
> ~/tensorflow/bazel-bin/tensorflow/contrib/lite/toco/toco \
--input_file=./RD_Net/RD_Net_opt.pb \
--output_file=./RD_Net.tflite \
--input_format=TENSORFLOW_GRAPHDEF \
--output_format=TFLITE \
--inference_type=FLOAT \
--input_data_types=FLOAT \
--input_arrays=Placeholder \
--output_arrays=seq_4/Conv/BiasAdd \
--input_shapes=1,240,320,3

### Describe the problem
I save my pre-trained model into .pb file according to [freeze_graph.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tools/freeze_graph.py) and [optimize_for_inference.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tools/optimize_for_inference.py)

my input/out node in .pbtxt

> node {
  name: "Placeholder"
  op: "Placeholder"
  attr {
    key: "dtype"
    value {
      type: DT_FLOAT
    }
  }
  attr {
    key: "shape"
    value {
      shape {
        dim {
          size: 1
        }
        dim {
          size: 240
        }
        dim {
          size: 320
        }
        dim {
          size: 3
        }
      }
    }
  }
}

> node {
  name: "seq_4/Conv/BiasAdd"
  op: "BiasAdd"
  input: "seq_4/Conv/Conv2D"
  input: "seq_4/Conv/biases"
  attr {
    key: "T"
    value {
      type: DT_FLOAT
    }
  }
  attr {
    key: "data_format"
    value {
      s: "NHWC"
    }
  }
}

I use the .pb to test prediction and it work fine. (the in/out node name and size are checked)
but fail in .tflite converting flow...  any suggestion?

My step:
1. clone last tensorflow (commit 72239e9b0432a26feadc2412abf89a8fb92828f0)

2. build the toco:
> bazel build tensorflow/contrib/lite/toco:toco

3. and convert my .pb into .tflite:
> ~/tensorflow/bazel-bin/tensorflow/contrib/lite/toco/toco \
--input_file=./RD_Net/RD_Net_opt.pb \
--output_file=./RD_Net.tflite \
--input_format=TENSORFLOW_GRAPHDEF \
--output_format=TFLITE \
--inference_type=FLOAT \
--input_data_types=FLOAT \
--input_arrays=Placeholder \
--output_arrays=seq_4/Conv/BiasAdd \
--input_shapes=1,240,320,3

the error msg:

> 2018-03-06 15:39:30.859538: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before Removing unused ops: 1365 operators, 2579 arrays (0 quantized)
2018-03-06 15:39:30.910650: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before general graph transformations: 1365 operators, 2579 arrays (0 quantized)
2018-03-06 15:39:30.967113: F tensorflow/contrib/lite/toco/graph_transformations/propagate_fixed_sizes.cc:764] Check failed: output_size_shape.dimensions_count() == 1 (2 vs. 1)
Aborted (core dumped)

**Comment by poxvoculi** [comment]
@andrehentz Can you take a look at this or recommend someone who works on this?

**Comment by andrehentz** [comment]
It looks like you have a resize_bilinear operations where the 'shape' is 2D but TOCO expects it to be 1D (which seems to match the documentation at https://www.tensorflow.org/api_docs/python/tf/image/resize_bilinear

**Comment by ChiFang** [comment]
Hi @andrehentz  how do you known have a resize_bilinear operations? the error msg doesn't show the key word. 
I do have resize_bilinear for upsample.
it is weird that if I use  graph_transforms tools by my self with cmd line after "optimize_for_inference.py", the converter work fine to get .tflite but fail at android app.

I finally change "resize_bilinear"  into "TransposeConv" but fail with #17501 

Any suggestion? Thanks~~~

**Comment by andrehentz** [comment]
I looke at tensorflow/contrib/lite/toco/graph_transformations/propagate_fixed_sizes.cc:764 which is handling resize bilinear. Why is the shape of your resize_bilinear 2D? Is that something that TF supports but TF Lite doesn't?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrehentz: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by dodiknikola** [comment]
Hi, I'm having the same issue. I'm using the frozen DeepLab v3+ model from [here](https://github.com/tensorflow/models/tree/master/research/deeplab) ([xception_coco_voc_trainaug](https://github.com/tensorflow/models/blob/master/research/deeplab/g3doc/model_zoo.md)).

I printed out the shapes of all `ResizeBilinear/size` tensors, as well as the shapes of the inputs to all of the `ResizeBilinear` tensors, and it seems that they are all 1D and have two elements:

```
In [59]: print([(n.name, n.outputs[0].get_shape(), tf.contrib.util.constant_value(n.outputs[0])) for n in tf.get_default_graph().get_operations() if 'Bilinear' in n.name and '/size' in n.name])

[('import/ResizeBilinear_1/size', TensorShape([Dimension(2)]), array([65, 65], dtype=int32)), ('import/decoder/ResizeBilinear/size', TensorShape([Dimension(2)]), array([129, 129], dtype=int32)), ('import/decoder/ResizeBilinear_1/size', TensorShape([Dimension(2)]), array([129, 129], dtype=int32)), ('import/ResizeBilinear_2/size', TensorShape([Dimension(2)]), array([129, 129], dtype=int32))]
```

```
In [60]: print([(n.name, n.inputs[1].name, n.inputs[1].shape, tf.contrib.util.constant_value(n.inputs[1])) for n in tf.get_default_graph().get_operations() if 'Bilinear' in n.name and '/size' not in n.name])

[('import/ResizeBilinear', 'import/ToInt32:0', TensorShape([Dimension(2)]), None), ('import/ResizeBilinear_1', 'import/ResizeBilinear_1/size:0', TensorShape([Dimension(2)]), array([65, 65], dtype=int32)), ('import/decoder/ResizeBilinear', 'import/decoder/ResizeBilinear/size:0', TensorShape([Dimension(2)]), array([129, 129], dtype=int32)), ('import/decoder/ResizeBilinear_1', 'import/decoder/ResizeBilinear_1/size:0', TensorShape([Dimension(2)]), array([129, 129], dtype=int32)), ('import/ResizeBilinear_2', 'import/ResizeBilinear_2/size:0', TensorShape([Dimension(2)]), array([129, 129], dtype=int32)), ('import/ResizeBilinear_3', 'import/strided_slice_7:0', TensorShape([Dimension(2)]), None)]
```

However, when I add the following code to `ProcessResizeBilinearOperator`,

```
  std::cout << "Output size shape for " << output_size_name << ": ";
  for(auto dim : output_size_shape.dims()) {
    std::cout << dim << ", ";
  }
  std::cout << std::endl;
  // CHECK_EQ(output_size_shape.dimensions_count(), 1);
```

I get the output:

```
2018-03-27 16:44:23.330779: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1236] Converting unsupported operation: Equal
2018-03-27 16:44:23.330937: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1236] Converting unsupported operation: LogicalAnd
2018-03-27 16:44:23.453760: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before Removing unused ops: 1895 operators, 3070 arrays (0 quantized)
2018-03-27 16:44:23.504983: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] After Removing unused ops pass 1: 1885 operators, 3051 arrays (0 quantized)
2018-03-27 16:44:23.573858: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before general graph transformations: 1885 operators, 3051 arrays (0 quantized)
Output size shape for ToInt32: 2, 1, 
2018-03-27 16:44:23.803197: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] After general graph transformations pass 1: 460 operators, 1136 arrays (0 quantized)
Output size shape for ToInt32: 2, 1, 
2018-03-27 16:44:23.817572: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before dequantization graph transformations: 460 operators, 1136 arrays (0 quantized)
2018-03-27 16:44:23.827669: I tensorflow/contrib/lite/toco/allocate_transient_arrays.cc:311] Total transient array allocated size: 7105728 bytes, theoretical optimal value: 6316160 bytes.
2018-03-27 16:44:23.830342: F tensorflow/contrib/lite/toco/tooling_util.cc:1818] Check failed: array.final_data_type == array.data_type Array "ImageTensor" has mis-matching actual and final data types (4,2).
./convert.sh: line 10: 30075 Aborted                 (core dumped) bazel-bin/tensorflow/contrib/lite/toco/toco --input_file=../deeplabv3/deeplabv3_pascal_train_aug/frozen_inference_graph.pb --input_format=TENSORF
LOW_GRAPHDEF --output_format=TFLITE --output_file=../deeplabv3/deeplabv3_pascal_train_aug/deeplabv3.tflite --inference_type=FLOAT --input_type=FLOAT --input_arrays=ImageTensor --output_arrays=SemanticPredictions 
--input_shapes=1,513,513,3

```

Does this information help?

**Comment by calicratis19** [comment]
I'm facing the same issue with mobilenet.

The command to generate tlite file

```
bazel-bin/tensorflow/contrib/lite/toco/toco \
> --input_file=frozen_inference_graph.pb \
> --input_format=TENSORFLOW_GRAPHDEF \
> --output_format=TFLITE \
> --output_file=frozen_inference_graph.lite \
> --inference_type=FLOAT \
> --input_type=FLOAT \
> --input_arrays=ImageTensor \
> --output_arrays=SemanticPredictions \
> --input_shapes=1,513,342,3

```
The log of the error.
```
2018-04-01 15:57:00.360279: W tensorflow/contrib/lite/toco/toco_cmdline_flags.cc:178] --input_type is deprecated. It was an ambiguous flag that set both --input_data_types and --inference_input_type. If you are trying to complement the input file with information about the type of input arrays, use --input_data_type. If you are trying to control the quantization/dequantization of real-numbers input arrays in the output file, use --inference_input_type.
2018-04-01 15:57:00.389245: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1172] Converting unsupported operation: Equal
2018-04-01 15:57:00.389504: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1172] Converting unsupported operation: LogicalAnd
2018-04-01 15:57:00.406372: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before Removing unused ops: 812 operators, 1241 arrays (0 quantized)
2018-04-01 15:57:00.434485: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] After Removing unused ops pass 1: 802 operators, 1222 arrays (0 quantized)
2018-04-01 15:57:00.465611: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before general graph transformations: 802 operators, 1222 arrays (0 quantized)
2018-04-01 15:57:00.465869: F tensorflow/contrib/lite/toco/graph_transformations/propagate_fixed_sizes.cc:737] Check failed: output_size_shape.dimensions_count() == 1 (2 vs. 1)
```

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrehentz: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by raninbowlalala** [comment]
same error as @calicratis19

**Comment by haroldyang977** [comment]
same error as @calicratis19

**Comment by ttyang1018** [comment]
same error as @calicratis19

**Comment by lizhaoda** [comment]
same error as @calicratis19

**Comment by tworuler** [comment]
same error as @calicratis19

**Comment by iChiaGuo** [comment]
same error as @calicratis19

**Comment by RoboEvangelist** [comment]
I get this error:
F tensorflow/contrib/lite/toco/tooling_util.cc:822] Check failed: d >= 1 (0 vs. 1)

**Comment by RoboEvangelist** [comment]
This error converting my LSTM graph:
[https://stackoverflow.com/questions/51275360/convert-lstm-graph-with-tflite-fails](https://stackoverflow.com/questions/51275360/convert-lstm-graph-with-tflite-fails)

**Comment by RoboEvangelist** [comment]
During inference, batch size = 1, 10 inputs, each input is of length 2560

**Comment by anitha-v** [comment]
Looks like DeepLab is converted successfully. For other issues please open separate bugs with additional information to help us debug this better.
